File: axon/runtime/esk/compliance.py
"""
AXON Runtime — Regulatory Compliance Registry
================================================
Canonical vocabulary for the ESK κ (regulatory class) annotation.

Every class in this registry must correspond to a real-world regulation
or framework; typos are compile-time errors thanks to §6.1 enforcement.

See docs/plan_io_cognitivo.md Fase 6.1 — "Compile-time Compliance".
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Iterable


# ═══════════════════════════════════════════════════════════════════
#  Registry
# ═══════════════════════════════════════════════════════════════════

@dataclass(frozen=True)
class RegulatoryClass:
    """Metadata for one regulatory framework."""
    name: str                        # canonical label (matches AST)
    title: str                       # human-readable name
    jurisdiction: str                # US | EU | Global | Sector
    sector: str                      # healthcare | financial | gov | ...
    description: str

    def covers(self, other: "RegulatoryClass") -> bool:
        """Reflexive coverage — a class covers itself.  Overlap between
        frameworks (e.g. HIPAA ⇒ some SOX obligations) is an explicit
        policy decision, not an implicit one."""
        return self.name == other.name


# Canonical set — must match type_checker._REGULATORY_CLASSES exactly.
REGISTRY: dict[str, RegulatoryClass] = {
    "HIPAA": RegulatoryClass(
        name="HIPAA",
        title="Health Insurance Portability and Accountability Act",
        jurisdiction="US",
        sector="healthcare",
        description="PHI confidentiality, integrity, and availability for US healthcare providers.",
    ),
    "PCI_DSS": RegulatoryClass(
        name="PCI_DSS",
        title="Payment Card Industry Data Security Standard",
        jurisdiction="Global",
        sector="financial",
        description="Cardholder data protection for merchants and payment processors.",
    ),
    "GDPR": RegulatoryClass(
        name="GDPR",
        title="General Data Protection Regulation",
        jurisdiction="EU",
        sector="cross-sector",
        description="Personal data protection for EU residents, with right-to-erasure.",
    ),
    "SOX": RegulatoryClass(
        name="SOX",
        title="Sarbanes-Oxley Act",
        jurisdiction="US",
        sector="financial",
        description="Financial reporting integrity for public companies.",
    ),
    "FINRA": RegulatoryClass(
        name="FINRA",
        title="Financial Industry Regulatory Authority",
        jurisdiction="US",
        sector="financial",
        description="Broker-dealer oversight — communications, record retention, surveillance.",
    ),
    "ISO27001": RegulatoryClass(
        name="ISO27001",
        title="ISO/IEC 27001",
        jurisdiction="Global",
        sector="cross-sector",
        description="Information security management system certification.",
    ),
    "SOC2": RegulatoryClass(
        name="SOC2",
        title="SOC 2 Type II",
        jurisdiction="Global",
        sector="cross-sector",
        description="Trust Services Criteria — security, availability, confidentiality.",
    ),
    "FISMA": RegulatoryClass(
        name="FISMA",
        title="Federal Information Security Management Act",
        jurisdiction="US",
        sector="government",
        description="US federal government information security baseline.",
    ),
    "GxP": RegulatoryClass(
        name="GxP",
        title="Good x Practice",
        jurisdiction="Global",
        sector="pharma",
        description="Quality guidelines for pharma / clinical / manufacturing (GLP/GMP/GCP).",
    ),
    "CCPA": RegulatoryClass(
        name="CCPA",
        title="California Consumer Privacy Act",
        jurisdiction="US-CA",
        sector="cross-sector",
        description="Consumer data rights for California residents.",
    ),
    "NIST_800_53": RegulatoryClass(
        name="NIST_800_53",
        title="NIST SP 800-53",
        jurisdiction="US",
        sector="government",
        description="Security and privacy controls catalog for US federal systems.",
    ),
}


# ═══════════════════════════════════════════════════════════════════
#  Coverage predicates — used by the compiler AND the runtime
# ═══════════════════════════════════════════════════════════════════

def is_known(label: str) -> bool:
    return label in REGISTRY


def get(label: str) -> RegulatoryClass:
    if label not in REGISTRY:
        raise KeyError(f"unknown regulatory class '{label}'")
    return REGISTRY[label]


def covers(shield_compliance: Iterable[str], required: Iterable[str]) -> set[str]:
    """Return the MISSING classes — i.e. required minus covered.

    A shield covers a required class exactly when the class appears in
    the shield's compliance list.  No transitive coverage: each class
    must be listed explicitly (audit trails demand this).
    """
    provided = set(shield_compliance or [])
    needed = set(required or [])
    return needed - provided


def classify_sector(labels: Iterable[str]) -> set[str]:
    """Return the set of regulated sectors the labels span — useful for
    dossiers and audit reporting."""
    sectors: set[str] = set()
    for label in labels:
        if label in REGISTRY:
            sectors.add(REGISTRY[label].sector)
    return sectors


__all__ = [
    "REGISTRY",
    "RegulatoryClass",
    "classify_sector",
    "covers",
    "get",
    "is_known",
]

File: axon-rs/src/esk/compliance.rs
//! AXON Runtime — Regulatory Compliance Registry (§Fase 6.1)
//!
//! Direct port of `axon/runtime/esk/compliance.py`.
//!
//! Canonical vocabulary for the ESK κ (regulatory class) annotation.
//! Every class corresponds to a real-world regulation; typos are
//! compile-time errors (§6.1 enforcement).

#![allow(dead_code)]

use std::collections::{HashMap, HashSet};

use serde::Serialize;

/// Metadata for one regulatory framework.
#[derive(Debug, Clone, Serialize)]
pub struct RegulatoryClass {
    pub name: String,
    pub title: String,
    pub jurisdiction: String,
    pub sector: String,
    pub description: String,
}

impl RegulatoryClass {
    /// Reflexive coverage — a class covers itself only. Cross-framework
    /// overlap is an explicit policy decision, not implicit.
    pub fn covers_class(&self, other: &RegulatoryClass) -> bool {
        self.name == other.name
    }

    fn new(
        name: &str,
        title: &str,
        jurisdiction: &str,
        sector: &str,
        description: &str,
    ) -> Self {
        RegulatoryClass {
            name: name.into(),
            title: title.into(),
            jurisdiction: jurisdiction.into(),
            sector: sector.into(),
            description: description.into(),
        }
    }
}

/// Build the canonical registry — must match Python `REGISTRY` exactly.
pub fn registry() -> HashMap<String, RegulatoryClass> {
    let entries = [
        RegulatoryClass::new(
            "HIPAA",
            "Health Insurance Portability and Accountability Act",
            "US",
            "healthcare",
            "PHI confidentiality, integrity, and availability for US healthcare providers.",
        ),
        RegulatoryClass::new(
            "PCI_DSS",
            "Payment Card Industry Data Security Standard",
            "Global",
            "financial",
            "Cardholder data protection for merchants and payment processors.",
        ),
        RegulatoryClass::new(
            "GDPR",
            "General Data Protection Regulation",
            "EU",
            "cross-sector",
            "Personal data protection for EU residents, with right-to-erasure.",
        ),
        RegulatoryClass::new(
            "SOX",
            "Sarbanes-Oxley Act",
            "US",
            "financial",
            "Financial reporting integrity for public companies.",
        ),
        RegulatoryClass::new(
            "FINRA",
            "Financial Industry Regulatory Authority",
            "US",
            "financial",
            "Broker-dealer oversight — communications, record retention, surveillance.",
        ),
        RegulatoryClass::new(
            "ISO27001",
            "ISO/IEC 27001",
            "Global",
            "cross-sector",
            "Information security management system certification.",
        ),
        RegulatoryClass::new(
            "SOC2",
            "SOC 2 Type II",
            "Global",
            "cross-sector",
            "Trust Services Criteria — security, availability, confidentiality.",
        ),
        RegulatoryClass::new(
            "FISMA",
            "Federal Information Security Management Act",
            "US",
            "government",
            "US federal government information security baseline.",
        ),
        RegulatoryClass::new(
            "GxP",
            "Good x Practice",
            "Global",
            "pharma",
            "Quality guidelines for pharma / clinical / manufacturing (GLP/GMP/GCP).",
        ),
        RegulatoryClass::new(
            "CCPA",
            "California Consumer Privacy Act",
            "US-CA",
            "cross-sector",
            "Consumer data rights for California residents.",
        ),
        RegulatoryClass::new(
            "NIST_800_53",
            "NIST SP 800-53",
            "US",
            "government",
            "Security and privacy controls catalog for US federal systems.",
        ),
    ];
    entries.into_iter().map(|c| (c.name.clone(), c)).collect()
}

pub fn is_known(label: &str) -> bool {
    registry().contains_key(label)
}

pub fn get_class(label: &str) -> Option<RegulatoryClass> {
    registry().get(label).cloned()
}

/// Return the MISSING classes — i.e. `required \ provided`.
pub fn covers<I, J, S>(shield_compliance: I, required: J) -> HashSet<String>
where
    I: IntoIterator<Item = S>,
    J: IntoIterator<Item = S>,
    S: AsRef<str>,
{
    let provided: HashSet<String> = shield_compliance
        .into_iter()
        .map(|s| s.as_ref().to_string())
        .collect();
    let needed: HashSet<String> = required
        .into_iter()
        .map(|s| s.as_ref().to_string())
        .collect();
    needed.difference(&provided).cloned().collect()
}

/// Return the set of sectors the labels span.
pub fn classify_sector<I, S>(labels: I) -> HashSet<String>
where
    I: IntoIterator<Item = S>,
    S: AsRef<str>,
{
    let reg = registry();
    labels
        .into_iter()
        .filter_map(|l| reg.get(l.as_ref()).map(|c| c.sector.clone()))
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn registry_has_all_expected_classes() {
        let reg = registry();
        for name in [
            "HIPAA", "PCI_DSS", "GDPR", "SOX", "FINRA", "ISO27001", "SOC2",
            "FISMA", "GxP", "CCPA", "NIST_800_53",
        ] {
            assert!(reg.contains_key(name), "missing {name}");
        }
    }

    #[test]
    fn is_known_accepts_canonical_names_only() {
        assert!(is_known("HIPAA"));
        assert!(!is_known("hipaa")); // case-sensitive
        assert!(!is_known("INVENTED"));
    }

    #[test]
    fn covers_returns_missing_classes() {
        let missing = covers(["HIPAA", "SOC2"], ["HIPAA", "GDPR", "SOC2"]);
        assert_eq!(missing.len(), 1);
        assert!(missing.contains("GDPR"));
    }

    #[test]
    fn classify_sector_aggregates() {
        let sectors = classify_sector(["HIPAA", "PCI_DSS", "GxP"]);
        assert!(sectors.contains("healthcare"));
        assert!(sectors.contains("financial"));
        assert!(sectors.contains("pharma"));
    }

    #[test]
    fn reflexive_covers_class() {
        let hipaa = get_class("HIPAA").unwrap();
        assert!(hipaa.covers_class(&hipaa));
        let gdpr = get_class("GDPR").unwrap();
        assert!(!hipaa.covers_class(&gdpr));
    }
}
